library/core/src/array/mod.rs

```diff
@@ -393,37 +395,125 @@ impl<T, const N: usize> [T; N] {
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(build_array)]
+///
+/// let array: [usize; 5] = core::array::build_array(|i| i);
+/// assert_eq!(array, [0, 1, 2, 3, 4]);
+/// ```
+#[inline]
+#[unstable(feature = "build_array", issue = "none")]
+pub fn build_array<F, T, const N: usize>(mut cb: F) -> [T; N]
+where
+    F: FnMut(usize) -> T,
+{
+    let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+    let mut guard: Guard<T, N> = Guard::new(MaybeUninit::first_ptr_mut(&mut array));
+    for (idx, value) in array.iter_mut().enumerate() {
+        value.write(cb(idx));
+        guard.initialized += 1;
+    }
+    // FIXME: Convert to crate::mem::transmute once it works with generics.
+    mem::forget(guard);
+    // SAFETY: It is guaranteed that the whole array will be filled
+    unsafe { mem::transmute_copy::<_, [T; N]>(&array) }
+}
+
+/// Creates an array `[T; N]` where each fallible array element `T` is returned by the `cb` call.
+/// Unlike `build_array`, where the element creation can't fail, this version will return an error
+/// if any element creation was unsuccessful.
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(build_array)]
+///
+/// use core::array::try_build_array;
+///
+/// #[derive(Debug, PartialEq)]
+/// enum SomeError {
+///     Foo
+/// }
+///
+/// let array: Result<[usize; 5], SomeError> = try_build_array(|i| Ok(i));
+/// assert_eq!(array, Ok([0, 1, 2, 3, 4]));
+///
+/// let another_array: Result<[usize; 5], SomeError> = try_build_array(|_| Err(SomeError::Foo));
+/// assert_eq!(another_array, Err(SomeError::Foo));
+/// ```
+
+#[inline]
+#[unstable(feature = "build_array", issue = "none")]
+pub fn try_build_array<E, F, T, const N: usize>(mut cb: F) -> Result<[T; N], E>
+where
+    F: FnMut(usize) -> Result<T, E>,
+{
+    let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+    let mut guard: Guard<T, N> = Guard::new(MaybeUninit::first_ptr_mut(&mut array));
+    for (idx, value) in array.iter_mut().enumerate() {
+        match cb(idx) {
+            Err(e) => {
+                // FIXME: Convert to crate::mem::transmute once it works with generics.
+                mem::forget(guard);
+                return Err(e);
+            }
+            Ok(rslt) => {
+                value.write(rslt);
+                guard.initialized += 1;
+            }
+        }
+    }
+    // FIXME: Convert to crate::mem::transmute once it works with generics.
+    mem::forget(guard);
+    // SAFETY: If every creation was successful, it is guaranteed that the whole array was filled.
+    Ok(unsafe { mem::transmute_copy::<_, [T; N]>(&array) })
+}
+
+struct Guard<T, const N: usize> {
+    dst: *mut T,
+    initialized: usize,
+}
```

Comment:
Ah there's also another PR for proposing a general purpose guard that also is responsible for filling an array(#75717) which might make this more ergonomic: `FillError::new((0..N).map(cb)).unwrap()`. I'm not sure whether it's worth waiting for that to land but just wanted to point it out

Reply:
This also got me thinking about unwind-safe initialization in the standard library recently too and sketched out a few general-purpose functions that would also fill this role: https://github.com/KodrAus/poison-guard/blob/main/src/guard.rs

Reply:
The PR was closed in favour of a possible std version of `ArrayVec` so it doesn't seem relevant anymore. Marking as resolved.

BTW, the `ArrayVec` RFC still didn't get an official response